hogben/simulate.py

```diff
@@ -10,368 +10,270 @@
+            probe.spin_state = spin_state
+
+        return refl1d.experiment.Experiment(probe=probe, sample=self.sample_model)
+
+    def reflectivity(self, q: np.ndarray) -> np.ndarray:
+        """Calculates the model reflectivity at given `q` points.
+
+        Args:
+            q: Q points to calculate reflectance at.
+
+        Returns:
+            numpy.ndarray: reflectivity for each Q point.
+
+        """
+        # If there are no data points, return an empty array.
+        if len(q) == 0:
+            return np.array([])
+
+        # Calculate the reflectance in either refnx or Refl1D.
+        if isinstance(self.sample_model, refnx.reflect.Stucture):
+            self.model = refnx.reflect.ReflectModel(self.sample_model,
+                                                    scale=self.scale,
+                                                    bkg=self.bkg,
+                                                    dq=self.dq)
+            return self.model(q)
+
+        if isinstance(self.sample_model, refl1d.model.Stack):
+            # If magnetic, use the correct spin state.
+            experiment = self.refl1d_experiment(q,
+                                                self.sample_model.probe.spin_state)
+
+            if self.sample_model.ismagnetic:
+                return experiment.reflectivity()[self.sample_model.probe.spin_state][1]
+            # experiment.reflectivity() returns q, r, or an array if magnetic
+
+            else:
+                return experiment.reflectivity()[1]
+
+    def _run_experiment(self, angle: float, points: int, time: float,
+                        spin_state: Optional[str] = None) -> tuple:
+        """Simulates a single angle measurement of a given 'model' on the
+        instrument set in self.incident_flux_data with the defined spin state,
+        unpolarised measurements (the default) have the set to None
+
+        Args:
+            angle: angle to simulate.
+            points: number of points to use for simulated data.
+            time: counting time for simulation.
+            spin_state: spin state to simulate if given a magnetic sample.
+
+        Returns:
+            tuple: simulated Q, R, dR data and incident neutron counts.
+
+        """
+        # direct_beam = [wavelength, flux]
+        wavelengths, flux = self.incident_flux_data
+
+        # Scale flux by relative measurement angle squared (assuming both slits
+        # scale linearly with angle, this should be correct)
+        scaled_flux = flux * pow(angle / self.angle_scale, 2)
+
+        q = 4 * np.pi * np.sin(np.radians(angle)) / wavelengths
+
+        # Bin Q's' in equally geometrically-spaced bins using flux as weighting
+        q_bin_edges = np.geomspace(min(q), max(q), points + 1)
+        flux_binned, _ = np.histogram(q, q_bin_edges, weights=scaled_flux)
+        # Calculate the number of incident neutrons for each bin.
+        counts_incident = flux_binned * time
```

Comment:
Not really a major issue, but I noticed that in the old code, counts_incident is assigned just before we actually need it. Here it's defined 20 lines early. Any specific reason for that?